### PR TITLE
use null aware operators

### PR DESCRIPTION
**Pull request by a14n**

### CODE REVIEW

File: packages/flutter/lib/src/gestures/arena.dart [hunk 1]
```diff
@@ -266,7 +266,7 @@ class GestureArenaManager {
   bool _debugLogDiagnostic(int pointer, String message, [ _GestureArena state ]) {
     assert(() {
       if (debugPrintGestureArenaDiagnostics) {
-        final int count = state != null ? state.members.length : null;
+        final int count = state?.members?.length;
```
Comment (dnfield) [inline packages/flutter/lib/src/gestures/arena.dart]: Isn't this more expensive now?  Before we only checked if `state` was null, now we have to check both `state` and `state.members` (and presumably, there's some contract saying `members` is never null).

File: packages/flutter/test/material/dropdown_test.dart [hunk 2]
```diff
@@ -63,13 +63,13 @@ Widget buildFrame({
...
             underline: underline,
-            items: items == null ? null : items.map<DropdownMenuItem<String>>((String item) {
+            items: items?.map<DropdownMenuItem<String>>((String item) {
               return DropdownMenuItem<String>(
                 key: ValueKey<String>(item),
                 value: item,
                 child: Text(item, key: ValueKey<String>(item + 'Text')),
               );
-            }).toList(),
+            })?.toList(),
```
Comment (dnfield) [inline packages/flutter/test/material/dropdown_test.dart]: Same thing here

File: packages/flutter/lib/src/painting/gradient.dart [hunk 3]
```diff
@@ -606,7 +606,7 @@ class RadialGradient extends Gradient {
       radius * rect.shortestSide,
       colors, _impliedStops(), tileMode,
       null, // transform
-      focal == null  ? null : focal.resolve(textDirection).withinRect(rect),
+      focal?.resolve(textDirection)?.withinRect(rect),
```
Comment (dnfield) [inline packages/flutter/lib/src/painting/gradient.dart]: ditto

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 4]
```diff
@@ -518,7 +518,7 @@ class _IOSSimulatorLogReader extends DeviceLogReader {
       onListen: _start,
       onCancel: _stop,
     );
-    _appName = app == null ? null : app.name.replaceAll('.app', '');
+    _appName = app?.name?.replaceAll('.app', '');
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: ditto

**Comment by dnfield** [comment]
Probably more important than adding an additional branch in some places for the additional null check is what this does for NNBD - does it help that or make it harder (because now a value is being checked for null that wasn't before and the compiler will think it can be null?).

**Comment by a14n** [comment]
You're right: `state?.members?.length` adds an additional branch compared to `state != null ? state.members.length : null`. I don't think the 4 occurances you point out will introduce a real change in perf.

Regarding NNBD my understanding is that `state != null ? state.members.length : null` could be changed to `state?.members.length` to get exactly the same behaviour (assuming `members` is not nullable). @leafpetersen could you confirm?

File: packages/flutter/lib/src/gestures/arena.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/gestures/arena.dart]: this is in an assert, performance doesn't matter

File: packages/flutter/lib/src/painting/gradient.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/lib/src/painting/gradient.dart]: i doubt this has much effect but it's probably worth benchmarking one of these cases to check.

File: packages/flutter/test/material/dropdown_test.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/test/material/dropdown_test.dart]: this one i think the original is more readable. It's very confusing to have the random question mark multiple lines down this expression.

**Comment by Hixie** [comment]
I prefer `a == null ? null : a.b.c` over `a?.b?.c` but mostly for readability and correctness reasons, rather than perf reasons -- it's not clear in the latter whether we're expecting `b` to return null, or just handling the case of `a.b`, and if we're not expecting `b` to ever return null but it does, we'll end up swallowing the error.

**Comment by a14n** [comment]
IIUC in the NNDB world `a?.b.c` will be exactly the same as `a == null ? null : a.b.c`. But today `a?.b.c` is the same as `(a == null ? null : a.b).c`. That why the `can_be_null_after_null_aware` lint requires a `?.c` to prevent NPE if `a` is null.

If my understanding of NNBD is correct we could provide a lint to prevent using `?.` on non-nullable type (or even better the spec will forbid to use `?.` on a non-nullable type). So `a?.b?.c` will be changed to `a?.b.c`.

I will rollback the 4 places pointed out and wait for NNBD to re-apply a change to get `a?.b.c`

**Comment by a14n** [comment]
PTAL

**Review by dnfield — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
> If my understanding of NNBD is correct we could provide a lint to prevent using `?.` on non-nullable type (or even better the spec will forbid to use `?.` on a non-nullable type).

Having the analyzer complain if you use `?` on a non-nullable type would be fantastic.

**Comment by leafpetersen** [comment]
@a14n 

> IIUC in the NNDB world `a?.b.c` will be exactly the same as `a == null ? null : a.b.c`. But today `a?.b.c` is the same as `(a == null ? null : a.b).c`. That why the `can_be_null_after_null_aware` lint requires a `?.c` to prevent NPE if `a` is null.

This is correct.

**Comment by leafpetersen** [comment]
> > If my understanding of NNBD is correct we could provide a lint to prevent using `?.` on non-nullable type (or even better the spec will forbid to use `?.` on a non-nullable type).
> 
> Having the analyzer complain if you use `?` on a non-nullable type would be fantastic.

We came down on the side of not specifying this as an error, since it makes turning a nullable return type into a non-nullable return type a breaking change.  But agreed that this is a great candidate for a lint.

cc @bwilkerson @pq @stereotype441

**Pull request merged**